samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,889 @@
+	struct http2_client_ctx *ctx_client, int client_index)
+{
+	const char *data;
+	int len;
+	int total_received = 0;
+	int offset = 0;
+
+	http_parser_init(&parser, HTTP_REQUEST);
+	http_parser_settings_init(&parserSettings);
+	parserSettings.on_header_field = on_header_field;
+	http_parser_execute(&parser, &parserSettings,
+		ctx_client->buffer + offset, ctx_client->offset);
+
+	total_received += ctx_client->offset;
+	offset += ctx_client->offset;
+
+	if (offset >= MAX_HTTP_RESPONSE_SIZE)
+		offset = 0;
+
+	if (has_upgrade_header == false) {
+		const char *response = "HTTP/1.1 101 Switching Protocols\r\n"
+				       "Connection: Upgrade\r\n"
+				       "Upgrade: h2c\r\n"
+				       "\r\n";
+		if (sendall(ctx_client->client_fd, response, strlen(response)) <
+			0)
+			close_client_connection(ctx_server, client_index);
+
+		memset(ctx_client->buffer, 0, sizeof(ctx_client->buffer));
+		ctx_client->offset = 0;
+		ctx_client->server_state = CLOSING;
+
+	} else {
+		data = content;
+		len = sizeof(content);
+
+		http_parser_init(&parser, HTTP_REQUEST);
+		http_parser_settings_init(&parserSettings);
+		parserSettings.on_url = on_url;
+		http_parser_execute(&parser, &parserSettings,
+			ctx_client->buffer, sizeof(ctx_client->buffer));
+
+		if (strcmp(url_buffer, "/") == 0) {
+			sprintf(http_response,
+				"HTTP/1.1 200 OK\r\n"
+				"Content-Type: text/html\r\n"
+				"Content-Encoding: gzip\r\n"
+				"Content-Length: %d\r\n\r\n",
+				len);
+			if (sendall(ctx_client->client_fd, http_response,
+				    strlen(http_response)) < 0) {
+				close_client_connection(
+					ctx_server, client_index);
+			} else {
+
+				if (sendall(ctx_client->client_fd, data, len) <
+					0) {
+					LOG_ERR("sendall failed");
+					close_client_connection(
+						ctx_server, client_index);
+				}
+			}
+		} else {
+			const char *not_found_response =
+				"HTTP/1.1 404 Not Found\r\n"
+				"Content-Length: 9\r\n\r\n"
+				"Not Found";
+			if (sendall(ctx_client->client_fd, not_found_response,
+				    strlen(not_found_response)) < 0) {
+				close_client_connection(
+					ctx_server, client_index);
+			}
+		}
+		close_client_connection(ctx_server, client_index);
+		memset(ctx_client->buffer, 0, sizeof(ctx_client->buffer));
+		ctx_client->offset = 0;
+		ctx_client->server_state = CLOSING;
+	}
+
+	return 0;
+}
+
+int handle_initial_state(struct http2_client_ctx *ctx_client)
+{
+	if (!ctx_client->is_http2) {
+		ctx_client->server_state = HTTP_PREFACE;
+	} else {
+		if (ctx_client->streams[0].stream_state == -1) {
+			ctx_client->server_state =
+				determine_server_state(ctx_client->buffer);
+		} else {
+			ctx_client->server_state = STREAMING;
+		}
+	}
+
+	return 0;
+}
+
+int handle_http2_frame_setting(struct http2_client_ctx *ctx_client)
+{
+	printf("HTTP2_FRAME_SETTINGS\n");
+	int frame_count = parse_http2_frames(
+		ctx_client->buffer, ctx_client->offset, frames);
+	print_http2_frames(frames, frame_count);
+
+	if (!is_first_flag_set(ctx_client->buffer[4])) {
```

Comment:
`first_flag` is not a good name - this flag was given a very specific name in the RFC, ACK. Also, we need to keep in mind, that different frame types define different flags, so we need to consider this to create an unique name (for example the same bit will be occupied by END_STREAM flag in the headers frame). Therefore for this particular case I think `settings_ack_flag` could be a good choice.